### PR TITLE
Added WithExpressionExpanding configuration extension for EF Core.

### PR DESCRIPTION
**Pull request by sdanyliv**
Well, added extensions which simplifies expanding EF Core queries. For now it is not required to call `AsExpandable()` but just registering with `DbContextOptionsBuilder` `WithExpressionExpanding` extension.
```cs
var builder = new DbContextOptionsBuilder<TestContext>();
builder
    .UseSqlite("DataSource=testdb;mode=memory;cache=shared")
    .WithExpressionExpanding();
```
And then you can create context which queries will be expanded automatically
```cs
var context = new TestContext(builder.Options);
```
This also work with Dependency Injection.

### CODE REVIEW

File: tests/LinqKit.Tests.Net452/PredicateBuilderTests.cs [hunk 1]
```diff
@@ -41,7 +41,7 @@ public void PredicateBuilder_New_expr()
             var predicate = PredicateBuilder.New(ienumerable, expr: x => false);
             predicate = predicate.Or(x => x.number <= 2);
             predicate = predicate.Or(x => x.squared >= 64);
-            Assert.Equal("x => ((x.number <= 2) OrElse (x.squared >= 64))", predicate.Expand().ToString());
+            Assert.Equal("x => ((False OrElse (x.number <= 2)) OrElse (x.squared >= 64))", predicate.Expand().ToString());
         }
```
Comment (sdanyliv) [inline tests/LinqKit.Tests.Net452/PredicateBuilderTests.cs:45]: Found this bad test and bug in this function.

**Comment by StefH** [comment]
@sdanyliv 
Can you please solve merge conflict?

**Comment by sdanyliv** [comment]
Done, I'm glad that you are back ;)

**Comment by StefH** [comment]
@sdanyliv 
Will this work for all EF versions?

**Comment by sdanyliv** [comment]
For EF Core 3 and 5, For EF Core 6 should be created new library.

**Comment by sdanyliv** [comment]
But maybe, it may work with EF Core 6 without any additional manipulations. Not tested.

File: src/LinqKit.Microsoft.EntityFrameworkCore/ExpandableDbContextOptionsBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+            var extension = GetOrCreateExtension(optionsBuilder);
+            ((IDbContextOptionsBuilderInfrastructure)optionsBuilder).AddOrUpdateExtension(extension);
+
+            return optionsBuilder;
+        }
+
+        private static ExpandableDbContextOptionsExtension GetOrCreateExtension(DbContextOptionsBuilder optionsBuilder)
+            => optionsBuilder.Options.FindExtension<ExpandableDbContextOptionsExtension>()
+               ?? new ExpandableDbContextOptionsExtension();
```
Comment (StefH) [inline src/LinqKit.Microsoft.EntityFrameworkCore/ExpandableDbContextOptionsBuilderExtensions.cs:32]: can you place this on previous line?

File: src/LinqKit.Microsoft.EntityFrameworkCore/ExpandableDbContextOptionsBuilderExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        /// <summary>
+        /// Replaces method calls with lambda expressions.
+        /// </summary>
+        /// <param name="optionsBuilder">The builder being used to configure the context.</param>
+        /// <returns>The options builder so that further configuration can be chained.</returns>
+        public static DbContextOptionsBuilder WithExpressionExpanding(this DbContextOptionsBuilder optionsBuilder)
+        {
+            if (optionsBuilder is null)
+                throw new ArgumentNullException(nameof(optionsBuilder));
```
Comment (StefH) [inline src/LinqKit.Microsoft.EntityFrameworkCore/ExpandableDbContextOptionsBuilderExtensions.cs:22]: surround with { and }

This is a general comment for some more code you've added.

File: src/LinqKit.Microsoft.EntityFrameworkCore/ExpandableDbContextOptionsExtension.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                if (descriptor.ServiceType != typeof(IQueryTranslationPreprocessorFactory))
+                    continue;
+                if (descriptor.ImplementationType is null)
+                    continue;
+
+                // Add Injectable factory for actual implementation
+                services[index] = new ServiceDescriptor(
+                    descriptor.ServiceType,
+                    typeof(ExpandableQueryTranslationPreprocessorFactory<>)
+                        .MakeGenericType(descriptor.ImplementationType),
```
Comment (StefH) [inline src/LinqKit.Microsoft.EntityFrameworkCore/ExpandableDbContextOptionsExtension.cs:33]: append this line to previous line

File: src/LinqKit.Microsoft.EntityFrameworkCore/ExpandableDbContextOptionsExtension.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            public ExtensionInfo(IDbContextOptionsExtension extension)
+                : base(extension)
+            {
+            }
+
+            public override bool IsDatabaseProvider
+                => false;
+
+            public override string LogFragment
+                => "LINQKitExpandable";
```
Comment (StefH) [inline src/LinqKit.Microsoft.EntityFrameworkCore/ExpandableDbContextOptionsExtension.cs:65]: maybe change this into private const string?

File: src/LinqKit.Microsoft.EntityFrameworkCore/ExpandableQueryTranslationPreprocessorOptions.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿namespace LinqKit
+{
+    internal class ExpandableQueryTranslationPreprocessorOptions
+    {
+        public ExpandableQueryTranslationPreprocessorOptions()
```
Comment (StefH) [inline src/LinqKit.Microsoft.EntityFrameworkCore/ExpandableQueryTranslationPreprocessorOptions.cs:5]: do you need an empty constructor?

File: tests/LinqKit.Microsoft.EntityFrameworkCore31.Tests/LinqKit.Microsoft.EntityFrameworkCore31.Tests.csproj [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    <GenerateAssemblyCompanyAttribute>false</GenerateAssemblyCompanyAttribute>
+    <GenerateAssemblyProductAttribute>false</GenerateAssemblyProductAttribute>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\LinqKit.Microsoft.EntityFrameworkCore31\LinqKit.Microsoft.EntityFrameworkCore31.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Compile Include="..\LinqKit.Microsoft.EntityFrameworkCore.TestsFiltes\**">
```
Comment (StefH) [inline tests/LinqKit.Microsoft.EntityFrameworkCore31.Tests/LinqKit.Microsoft.EntityFrameworkCore31.Tests.csproj:19]: `TestsFiltes` ?

**Review by StefH — CHANGES_REQUESTED** [review]

**Comment by sdanyliv** [comment]
Done

**Review by StefH — APPROVED** [review]

**Comment by rizi** [comment]
@sdanyliv will this work when Include is used?
At the moment when I don't use WithExprrssionExpanding, I have to call Include before AsExpandable?
Br

**Pull request merged**